src/java.base/share/classes/sun/security/util/Cache.java

```diff
@@ -360,26 +392,64 @@ public synchronized void clear() {
             for (CacheEntry<K,V> entry : cacheMap.values()) {
                 entry.invalidate();
             }
-            while (queue.poll() != null) {
-                // empty
-            }
+            while (queue.poll() != null);
         }
         cacheMap.clear();
     }
 
-    public synchronized void put(K key, V value) {
+    public void put(K key, V value) {
+        put(key, value, false);
+    }
+
+    /**
+     * This puts an entry into the cacheMap.
+     *
+     * If useQueue is true, V will be added using a QueueCacheEntry which
+     * is added to cacheMap.  If false, V is added to the cacheMap directly.
+     *
+     * This method is synchronized to avoid multiple QueueCacheEntry
+     * overwriting the same key.
+     *
+     * @param key key to the cacheMap
+     * @param value value to be stored
+     * @param canQueue can the value be put into a QueueCacheEntry
+     */
+    public synchronized void put(K key, V value, boolean canQueue) {
```

Comment:
In practice, can `QueueCacheEntry` and other `CacheEntry` instances be stored in a single cache?
If not, the argument `canQueue` in this method could be moved to class `MemoryCache` as a field.
Then, if `canQueue` is true, the `put` method always adds the value into the `QueueCacheEntry` associated with the `key`; otherwise, this method just uses other `CacheEntry` types.